### ISSUE TITLE
[K8S] DaemonSet causes remember me to not be honored (4.19+)

### ISSUE THREAD

**Issue by james-d-elliott** [authelia/authelia#1059]
The upgrade to the session provider in 4.19 causes a strange interaction only when Authelia is running in multiple pods (using redis). The remember me duration is not honored, instead the cookie expires based on the session expiration value.

It is confirmed that this interaction only happens with multiple pod architecture like a deployment with multiple pods or a daemonset replicated on all worker nodes (more than one).

**Comment by james-d-elliott** [comment]
Changing this to possible since it's not confirmed as a bug with Authelia directly, it may be upstream.

**Comment by james-d-elliott** [comment]
Resolved upstream.

**Comment by ikaruswill** [comment]
Hey @james-d-elliott, what was the fix eventually? Could you link the PR/Issue?

**Comment by james-d-elliott** [comment]
> 
> 
> Hey @james-d-elliott, what was the fix eventually? Could you link the PR/Issue? I've been facing this issue and following this thread but never got down to really deep diving on it.

@ikaruswill  I was just going back through old issues and while developing the helm chart I noted that the issue seems to have gone. Do you still have the issue on the latest version? I have yet to look at the changes upstream but I assumed that it got fixed there, Since it occurred when upgrading the session provider it made sense. I'll reopen until we figure out your issue.

**Comment by ikaruswill** [comment]
Thanks @james-d-elliott I'm using plain yaml manifests with my setup, and a deployment with multiple replicas instead of a daemonset.

Since you mentioned the helm chart, I'm thinking perhaps there might be some clues as to what is different between my setup and the helm chart's. Let me take a look there, and I'll report back on my findings.

**Comment by ikaruswill** [comment]
Wait a sec. Which chart did you use? I realize there's no official one.

**Comment by james-d-elliott** [comment]
I'm developing one currently. Though it shouldn't matter as long as it's a DaemonSet/Deployment using redis. Are you getting the behavior with both a DaemonSet and the Deployment with multiple replicas? Let me just double check my configuration isn't wonky.

**Comment by james-d-elliott** [comment]
> Wait a sec. Which chart did you use? I realize there's no official one.

I have a small workaround depending on which ingress provider you're using. I'm using the TraefikCRD ingress. Though there are probably solutions for each of them. However there definitely is still an issue on our end, going to experiment reverting to fasthttp/session [1.1.8](https://github.com/fasthttp/session/releases/tag/v1.1.8) in a test branch to see if it fixes it.

From there we can determine the best way to fix it. I'm also going to attempt to come up with a docker compose that can demonstrate the issue in a non-k8s environment so it's easy to test.

**Comment by james-d-elliott** [comment]
Think I'll mention you @ikaruswill just in case you're only subscribed to updates with mentions (not to be pushy).

**Comment by ikaruswill** [comment]
Yep I'm using a Deployment with 3 replicas, with 1 Redis as a session provider, also using Traefik IngressRoute CRDs. 

I'm cerrain it's not caused by the Redis being terminated as the age of the Redis pod is older than my session was when it expired with "remember me".

**Comment by james-d-elliott** [comment]
Using the following spec for the IngressRoute seems like a temporary workaround:

```yaml
spec:
  entryPoints:
  - websecure
  routes:
  - kind: Rule
    match: Host(`auth.example.com`) && PathPrefix(`/`)
    middlewares:
    - name: authelia
      namespace: authelia
    priority: 1
    services:
    - name: authelia
      port: 80
      sticky:
        cookie:
          httpOnly: true
          name: traefik-lb
          sameSite: none
          secure: true
      strategy: RoundRobin
```

**Comment by james-d-elliott** [comment]
Oh also thanks, Merry Christmas to you too! By the way here is the in-progress helm chart: https://github.com/authelia/chartrepo/tree/master/charts/experimental

**Comment by ikaruswill** [comment]
Thanks for the workaround and helm chart. 

Based on your workaround of sticky sessions, it seems like the problem lies with connections being routed to a different Authelia instance than what was initially authenticated through, and somehow the Session provider is not sharing the state with multiple Authelia replicas. 

What this also entails is that the probability of hitting session expiry increases with more replicas of Authelia running. Interesting..

I'll experiment with pure round-robin load-balancing to see if this completely breaks session persistence. This should prove our hypothesis..

**Comment by ikaruswill** [comment]
If I'm not wrong I believe I checked a few months back that on termination of the 'correct' Authelia pod, the session is lost. Let me verify that once again on latest.

**Comment by james-d-elliott** [comment]
Looks like you're correct, seems to me that as soon as my Traefik sticky cookie expires and it randomly chooses a new pod expires my session.

**Comment by james-d-elliott** [comment]
I believe I've fixed it. Will be able to test it soon with `authelia/authelia:bugfix-session-expiration-key` if it builds. Basically it looks like the session provider we're using was storing the session expiration key in a unique value per instance. I've temporarily forked it until they fix or advise on how we can work around it.

**Comment by ikaruswill** [comment]
That's pretty awesome quick work there @james-d-elliott . I'm glad we were able to pin this down quickly. This should bring Authelia a little closer to HA deployment! ☺️

I wonder why it doesn't seem to happen to others, they must be either on Docker Swarm/compose or K8s but have not realized it as they don't deploy multiple replicas of Authelia? 

Just food for thought.

**Comment by james-d-elliott** [comment]
This is now fixed in authelia/authelia:master - to be published officially with 4.25.0.
